### PR TITLE
Allow php8

### PR DESCRIPTION
**Pull request by fezfez**

### CODE REVIEW

**Comment by fezfez** [comment]
@PowerKiKi : error not related

File: composer.json [hunk 1]
```diff
@@ -33,7 +33,7 @@
         "friendsofphp/php-cs-fixer": "@stable",
         "laminas/laminas-diactoros": "@stable",
         "phpstan/phpstan": "@stable",
-        "phpunit/phpunit": "@stable"
+        "phpunit/phpunit": "^8.5||^9.2"
```
Comment (PowerKiKi) [inline composer.json]: This must not change, because it already allow to install 8 or 9.

```suggestion
        "phpunit/phpunit": "@stable"
```

File: composer.json [hunk 2]
```diff
@@ -24,7 +24,7 @@
         }
     },
     "require": {
-        "php": "^7.2",
+        "php": "^7.2||^8.0",
```
Comment (PowerKiKi) [inline composer.json]:
```suggestion
        "php": "^7.2 || ^8.0",
```

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -30,7 +31,7 @@ jobs:
           restore-keys: ${{ runner.os }}-composer-
 
       - name: Install dependencies
-        run: composer install
+        run: composer update
```
Comment (PowerKiKi) [inline .github/workflows/main.yml]: This must no change. Instead add a new step before this one that removes `composer.lock`, but only for PHP 8.0. Otherwise our dependencies are ever changing and our tests becomes unpredictable (as already seen in the error reported).

```suggestion
        run: composer install
```

File: composer.lock [hunk 4]
```diff
@@ -241,257 +241,33 @@
         }
     ],
     "packages-dev": [
-        {
```
Comment (PowerKiKi) [inline composer.lock]: The whole lock file must not change, because there is no need to update dependencies for this purpose. Instead after modifying the json file, use `composer update --lock` to prevent composer warnings while keeping existing dependencies.

**Review by PowerKiKi — CHANGES_REQUESTED** [review]
Thanks for your contribution. I added a few comments that must be resolved before merging.

File: .github/workflows/main.yml (hunk 3 above)
Comment (fezfez) [reply .github/workflows/main.yml]: A better way to check if this projects work with his declared dependencies would be to have this king of tests strategy https://github.com/Ocramius/DoctrineBatchUtils/blob/2.2.x/.github/workflows/phpunit.yml

Are you agree with this ?

**Comment by mfn** [comment]
There is a bit of chicken-egg problem with the current attempt here:
- project has a `composer.lock` file, currently with phpunit 8.5
- 8.5 is not compatible with PHP8
  at least _not yet_, see https://github.com/sebastianbergmann/phpunit/issues/4533
- hence the tests fails

Thus the initial idea to use `^8 || ^9` for phpunit makes sense but is torpedoed due to the `composer.lock`, otherwise the `composer update` step would be able to pick the right version itelf

I rarely see **libraries** having `composer.lock` exactly for this kind of reasons. Can it be removed?

It's only relevant for developing this library and is not respected when `graphql-upload` is used dependency. And in fact this circumstance seems to create a small issue here.

Just an observation, but maybe I got it backwards!

File: .github/workflows/main.yml (hunk 3 above)
Comment (PowerKiKi) [reply .github/workflows/main.yml]: If you refer to "lowest", "highest", "locked", then I don't want them for this project. I agree that it can have some values for some lib. But not here, because we virtually have a single dependency that really matters. The rest of dependencies is only used for developers and there is no point to test all those variations for a development environnement.

tl;dr; it's overkill for this lib

**Comment by PowerKiKi** [comment]
Thank you for mentioning the future compatibility of PHPUnit 8 and PHP 8. This is great news and will save a few headaches to many libs.

> Can it be removed ?

No, it must not. I am aware that some libraries choose not to commit composer.lock file. But IMHO this has huge downsides, such as introducing unpredictability in CI environnements. This is is clearly illustrated in the first version of this PR where PHP-Cs-Fixer "magically" got upgraded and [broke the CI](https://github.com/Ecodev/graphql-upload/pull/14/checks?check_run_id=1472540810#step:6:8).

While some communities are willing and have the resources to track a moving target, I am not and I don't. Keeping the lock is the cheapest solution.

> only relevant for developing

As you noted yourself, this has no incidence whatsoever on consuming users, and is only here to help developers and CI not waste time solving "syncing" issue.


Interestingly PHPUnit fails here because somehow [composer installs phpunit 4.4.5](https://github.com/Ecodev/graphql-upload/pull/14/checks?check_run_id=1472540989#step:5:79) (!) :roll_eyes:

**Comment by fezfez** [comment]
> Interestingly PHPUnit fails here because somehow composer installs phpunit 4.4.5 (!) roll_eyes

Ok so maybe we need to do a composer update with prefer lowest ? (in an other branch)

**Comment by PowerKiKi** [comment]
I believe the [solution I suggested](https://github.com/Ecodev/graphql-upload/pull/14/files#r532292227) is the simplest (until https://github.com/sebastianbergmann/phpunit/issues/4533 is released).

**Comment by PowerKiKi** [comment]
I had a better look and turns out that PHP-Cs-Fixer does not support PHP 8 and prevent us from installing a recent PHPUnit (so we end up with PHPUnit 4.4.5). So until https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4702 is solved, we need to remove the lock and remove this dependency, but only for PHP 8. Then we can `composer install` as usual, and the rest will pass.

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -31,6 +32,11 @@ jobs:
 
       - name: Install dependencies
         run: composer install
+        if: matrix.php-version != '8.0'
+
+      - name: Update dependencies
+        run: rm composer.lock && composer update
+        if: matrix.php-version == '8.0'
```
Comment (mfn) [inline .github/workflows/main.yml:42]: As mentioned in https://github.com/Ecodev/graphql-upload/pull/14#issuecomment-735770619 , the dependency on php-cs-fixer also needs to be removed, e.g.
`composer remove --dev --no-update friendsofphp/php-cs-fixer`

Hopefully it works then 🤞

**Comment by fezfez** [comment]
@PowerKiKi working and rebased

**Comment by mfn** [comment]
`\o/` ✅

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -31,6 +32,11 @@ jobs:
 
       - name: Install dependencies
         run: composer install
+        if: matrix.php-version != '8.0'
+
+      - name: Update dependencies
+        run: rm composer.lock && composer remove --dev --no-update friendsofphp/php-cs-fixer && composer update
```
Comment (mfn) [inline .github/workflows/main.yml]: Just FYI, it's often considered more readable to use multiple lines; otherwise it might also be easy to miss that there are 3 commands in total being executed:
```yaml
run: |
  rm composer.lock
  composer remove --dev --no-update friendsofphp/php-cs-fixer
  composer update
```
This works quite well without `&&` because github runs actions with `bash -e` which immediately exits on a failing command, i.e. if `composer remove…` would fail for some reason, the behaviour would be the same with the long line and `&&`

**Comment by PowerKiKi** [comment]
Much better thank you. The last thing we can improve is always run `composer install` and never runs `composer update`. I was going to do it myself, but it seems you did not allow edit on your PR. So I guess you'll have to do this:


```diff
diff --git .github/workflows/main.yml .github/workflows/main.yml
index ac52d08..3306e9e 100644
--- .github/workflows/main.yml
+++ .github/workflows/main.yml
@@ -30,16 +30,14 @@ jobs:
           key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.lock') }}
           restore-keys: ${{ runner.os }}-composer-
 
-      - name: Install dependencies
-        run: composer install
-        if: matrix.php-version != '8.0'
-
       - name: Update dependencies
+        if: matrix.php-version == '8.0'
         run: |
           rm composer.lock
           composer remove --dev --no-update friendsofphp/php-cs-fixer
-          composer update
-        if: matrix.php-version == '8.0'
+
+      - name: Install dependencies
+        run: composer install
 
       - name: Setup problem matchers for PHP
         run: echo "::add-matcher::${{ runner.tool_cache }}/php.json"
```

**Comment by fezfez** [comment]
> Much better thank you. The last thing we can improve is always run `composer install` and never runs `composer update`. I was going to do it myself, but it seems you did not allow edit on your PR. So I guess you'll have to do this:
> 
> ```diff
> diff --git .github/workflows/main.yml .github/workflows/main.yml
> index ac52d08..3306e9e 100644
> --- .github/workflows/main.yml
> +++ .github/workflows/main.yml
> @@ -30,16 +30,14 @@ jobs:
>            key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.lock') }}
>            restore-keys: ${{ runner.os }}-composer-
>  
> -      - name: Install dependencies
> -        run: composer install
> -        if: matrix.php-version != '8.0'
> -
>        - name: Update dependencies
> +        if: matrix.php-version == '8.0'
>          run: |
>            rm composer.lock
>            composer remove --dev --no-update friendsofphp/php-cs-fixer
> -          composer update
> -        if: matrix.php-version == '8.0'
> +
> +      - name: Install dependencies
> +        run: composer install
>  
>        - name: Setup problem matchers for PHP
>          run: echo "::add-matcher::${{ runner.tool_cache }}/php.json"
> ```

phpunit 4 is not compatible with php 8

**Comment by PowerKiKi** [comment]
Thank you for your contribution, I merged it with the additional modification I mentioned.
